### PR TITLE
Conseil #183/#184  query for distinct values given an attribute, and attribute filtering

### PR DESCRIPTION
**Pull request by piotrkosecki**
New set of features for metadata endpoint:
Endpoint for listing distinct values of given attributes - doesn't work for numeric and datetime fields as requested in the ticket - returns list of strings:
`{host}:{port}/metadata/networks/{network-name}/entities/{entity-name}/attributes/{attribute-name}`

Endpoint for listing distinct values of given attributes with filter:
`{host}:{port}/metadata/networks/{network-name}/entities/{entity-name}/attributes/{attribute-name}/filter/{filter-string}`

For security reasons `{filter-string}` can only contain letters and digits. We can extend it as we agree on some other safe solution.

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala [hunk 1]
```diff
@@ -88,4 +93,72 @@ object PlatformDiscoveryOperations {
...
+    ApiOperations.runQuerySequence(res).map(_.flatten)
+  }
+
+  /** Makes list of DBIO actions to get possible string values of the attributes
+    * @param  tableName name of the table from which we extract attributes
+    * @param  attribute name of the attribute
+    * @param  withFilter optional parameter which can filter attributes
+    * @return list of DBIO actions to get matching attributes
+    **/
+  def verifyAttributesAndGetQueries(tableName: String, attribute: String, withFilter: Option[String])
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: Is it feasible here to return a `DBIO[List[String]]` by `DBIO.sequence`-ing the `for` result and then `.map(_.flatten)` the ensuing `DBIO[List[List[String]]`?

For me it would make more sense to present a simpler type to the caller, instead of forcing him to flatten himself... unless the single operations in the list are to be used individually somewhere else...

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala [hunk 2]
```diff
@@ -88,4 +93,72 @@ object PlatformDiscoveryOperations {
...
+
+  /** Checks the data types if cannot be queried by */
+  private def canQueryType(dt: DataType): Boolean = {
+    // values described in the ticket #183
+    val cantQuery = List(DataType.Date, DataType.DateTime, DataType.Int, DataType.LargeInt, DataType.Decimal)
+    !cantQuery.contains(dt)
+  }
+
+  /** Checks if cardinality of the column is to high so it should not be queried */
+  private def isHighCardinality(distinctCount: Int): Boolean = {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: What about reversing the check and name this as an acceptance function (e.g. `isAcceptableColumnCardinality`, `isLimitedCardinality`, `isLowCardinality`...)?

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala [hunk 3]
```diff
@@ -88,4 +93,72 @@ object PlatformDiscoveryOperations {
...
+        case None =>
+          TezosDatabaseOperations.selectDistinct(tableName, column.name)
+      }
+    } yield distinctSelect
+  }
+
+  /** Checks the data types if cannot be queried by */
+  private def canQueryType(dt: DataType): Boolean = {
+    // values described in the ticket #183
+    val cantQuery = List(DataType.Date, DataType.DateTime, DataType.Int, DataType.LargeInt, DataType.Decimal)
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: `Set` is usually a better choice for inclusion-checks. It also allows to apply it as `!cantQuery(dt)`

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala [hunk 4]
```diff
@@ -88,4 +93,72 @@ object PlatformDiscoveryOperations {
...
+    !cantQuery.contains(dt)
+  }
+
+  /** Checks if cardinality of the column is to high so it should not be queried */
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: typo: "_to_ high"

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala [hunk 5]
```diff
@@ -45,20 +46,20 @@ object PlatformDiscoveryTypes {
...
   /**
-    * Function maps slick column type as a string to DataType
-    * @param tpe type as a string from slick TableQuery
+    * Function maps slick jdbc column type to DataType
+    * @param tpe type from slick TableQuery
     * @return one of DataType enums
     */
-  def mapType(tpe: String): DataType = {
-    val optionRegex = "Option\\[([A-Za-z0-9']+)\\]".r
+  def mapType(tpe: Type): DataType = {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala]: This is a great improvement

Will it create problems if we need to use any other db than postgres? Say, H2 for testing?

File: src/main/scala/tech/cryptonomic/conseil/util/RouteHandling.scala [hunk 6]
```diff
@@ -34,6 +34,13 @@ trait RouteHandling {
       case None => StatusCodes.NotFound
     }
 
+  protected def handleNoneAsTooLarge[T, R: ToResponseMarshaller](operation: => Future[Option[T]], converter: T => R = toJson[T] _)
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/util/RouteHandling.scala]: I can't seem to find if this is used somewhere

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 7]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
     }
   }
 
+  "listAttributeValues" should {
+
+    "return list of values of kind attribute of Fees without filter" in {
+      val avgFee = AverageFees(1, 3, 5, Timestamp.valueOf(LocalDateTime.of(2018, 11, 22, 12, 30)), "example1")
+
+      Await.result(dbHandler.run(TezosDatabaseOperations.writeFees(List(avgFee))), Duration.Inf)
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: Waiting an infinite amount of time doesn't seem useful, and may create issues on a CI environment.

Why not defining a reasonably low expectation instead, so that a problematic test will fail faster?

Side note: if you prefer there's also an extension method `future.isReadyWithin(Span)`that returns a Boolean, without using `Await.ready`

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 8]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
...
+      dbHandler.run(
+        DBIO.sequence(
+          PlatformDiscoveryOperations.verifyAttributesAndGetQueries("fees", "kind", None)
+        )
+      ).futureValue.flatten shouldBe List("example1")
+    }
+
+    "return throw an exeption when asked for medium attribute of Fees without filter - numeric attributes should not be displayed" in {
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: typo: "return throw an exception when..." -> "throw an exception when..."

personally I'd say that it "returns a failed future when..."

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 9]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
...
+      Await.result(dbHandler.run(TezosDatabaseOperations.writeFees(List(avgFee))), Duration.Inf)
+
+      val thrown = intercept[Exception] {
+        dbHandler.run(
+          DBIO.sequence(
+            PlatformDiscoveryOperations.verifyAttributesAndGetQueries("fees", "medium", None)
+          )
+        ).futureValue
+      }
+      thrown.getMessage shouldBe "The future returned an exception of type: java.util.NoSuchElementException, with message: Action.withFilter failed."
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: I guess there should be a way to extract a failed future's Exception type and check that without depending on a specific string message

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 10]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
...
+      dbHandler.run(
+        DBIO.sequence(
+          PlatformDiscoveryOperations.verifyAttributesAndGetQueries("fees", "kind", maliciousFilter)
+        )
+      ).futureValue.flatten shouldBe List.empty
+
+      dbHandler.run(Tables.Fees.length.result).futureValue shouldBe 1
+
+    }
+    "check if filtering works correctly" in {
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: This might be better phrased, the subject is "listAttributeValues"

The test case could read like, e.g. "listAttributeValues" should "correctly apply the filter" in ...

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 11]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
...
+      val avgFees = List(
+        AverageFees(1, 3, 5, Timestamp.valueOf(LocalDateTime.of(2018, 11, 22, 12, 30)), "example1"),
+        AverageFees(2, 4, 6, Timestamp.valueOf(LocalDateTime.of(2018, 11, 22, 12, 31)), "example2")
+      )
+      Await.result(dbHandler.run(TezosDatabaseOperations.writeFees(avgFees)), Duration.Inf)
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: Waiting infinite time, refer to previous comment

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 12]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
...
+
+      dbHandler.run(
+        DBIO.sequence(
+          PlatformDiscoveryOperations.verifyAttributesAndGetQueries("fees", "kind", None)
+        )
+      ).futureValue.flatten.size shouldBe 2
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: Why not test the exact values?

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala [hunk 13]
```diff
@@ -163,4 +171,83 @@ class PlatformDiscoveryOperationsTest
...
+      dbHandler.run(
+        DBIO.sequence(
+          PlatformDiscoveryOperations.verifyAttributesAndGetQueries("fees", "kind", Some("1"))
+        )
+      ).futureValue.flatten shouldBe List("example1")
+
+    }
+
+  }
+
```
Comment (ivanopagano) [inline src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: What about adding the edge case where no value is available in the DB?

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala (hunk 5 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala]: if we introduce more complex types it may, but for the generated ones I don't think so

Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala]: Are you sure that the postgres-profile-specific import won't create issues with other slick profiles? I was referring to that essentially

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala (hunk 1 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: ok, I'll change it but my reasoning was that I didn't want to repeat DBIO.sequence

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 9 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: I can only catch `org.scalatest.exceptions.TestFailedException` from what I see which hides my expected exception, so I can check for this exception but I would still need to check for the message - I can change to checking for `TestFailedException`- WDYT?

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala (hunk 2 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: ok, changed

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala (hunk 3 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: sure, changed

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala (hunk 4 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: fixed

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 7 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: yes, you're right, changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 8 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 10 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 11 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: done

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 12 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: changed

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 13 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: ok, added

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 9 above)
Comment (ivanopagano) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: You can always call `future.failed.futureValue` to get back the exception that failed the future.

Then you check the type with `should be a[NoSuchElementException]`

At least... I guess it should work

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala (hunk 1 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperations.scala]: I thought about that too, but personally I don't think this should be a driving factor behind code design choices. I tend to focus more on the intent and ergonomics of the resulting interface.

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala (hunk 5 above)
Comment (piotrkosecki) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala]: there are tests in place which are working fine - the whole schema is tested just because of this case. I think we should probably think abut using tests with the same database as we are using in production i.e. `postgresql-embedded` or with some docker containers.

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 9 above)
Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: with `future.failed.futureValue` it does not throw any exception:
`Expected exception java.util.NoSuchElementException to be thrown, but no exception was thrown`.

Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: ok, I found `future.value.get` which returns exception it supposed to

Comment (ivanopagano) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: `val ex = future.failed.futureValue` is supposed to be the exception itself, not to throw the exception... 😀

File: src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala (hunk 5 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryTypes.scala]: As long as tests work it's all fine.

About using embedded pg, it might make sense if it's fast. Using docker is ok for me, but only in integration tests. In my experience it gets burdensome on a daily working cycle. Unit tests are supposed to be slick

File: src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala (hunk 13 above)
Comment (ivanopagano) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: Oops, I meant a case when you have no rows in the database to start from

Comment (piotrkosecki) [reply src/test/scala/tech/cryptonomic/conseil/tezos/PlatformDiscoveryOperationsTest.scala]: Yes, I've added it before insert
